group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeSet.java

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * A {@code RangeSet} represents a range of integers from {@code from} (inclusive)
+ * to {@code to} (exclusive).
+ * This implementation provides a view over a continuous range of integers without actually storing them.
+ */
+public class RangeSet implements Set<Integer> {
+    private final int from;
+    private final int to;
+
+    /**
+     * Constructs a {@code RangeSet} with the specified range.
+     *
+     * @param from      The starting value (inclusive) of the range.
+     * @param to        The ending value (exclusive) of the range.
+     */
+    public RangeSet(int from, int to) {
+        this.from = from;
+        this.to = to;
+    }
+
+    @Override
+    public int size() {
+        return to - from;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (o instanceof Integer) {
+            int value = (Integer) o;
+            return value >= from && value < to;
+        }
+        return false;
+    }
+
+    @Override
+    public Iterator<Integer> iterator() {
+        return new Iterator<Integer>() {
+            private int current = from;
+
+            @Override
+            public boolean hasNext() {
+                return current < to;
+            }
+
+            @Override
+            public Integer next() {
+                if (!hasNext()) throw new NoSuchElementException();
+                return current++;
+            }
+        };
+    }
+
+    @Override
+    public Object[] toArray() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean add(Integer integer) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        for (Object o : c) {
+            if (!contains(o)) return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends Integer> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> c) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[");
+        for (int i = from; i < to; i++) {
+            sb.append(i);
+            if (i < to - 1) {
+                sb.append(", ");
+            }
+        }
+        sb.append("]");
+        return sb.toString();
```

Comment:
I wonder if we could just return something like `RangeSet(from=XX (inclusive), to=YY (exclusive))`. It is a tad annoying to compute the set to print it.

Reply:
I wanted it to look exactly like a set when we printed it, helps with comparisons in debugging at least, what do you think?

Reply:
The short version seems better to me but it may be a personal preference. I’d also like to have the type printed out.

Reply:
Okie I changed it!